### PR TITLE
Reduced time taken to compute expected proposers

### PR DESCRIPTION
**Pull request by rolfyone**
Tested on a mainnet state upgraded to capella, ran performance benchmarking on the first selection of state which will initially involve rejecting over 100k validators based on the fact they're not containing eth1 withdrawal credentials.

 - only compare credentials once
 - use single byte compare rather than slicing the string
 - don't touch the balance until the validator is known to have eth1 credentials, as that is surprisingly not a cheap operation. This could get flipped in time, but initially no validators are 0 balance and a lot of validators are not eth1 withdrawals.

 I shelved the benchmark, as it was based on a very specific mainnet state that would be sub-optimal to check in (very large)

before changes:
```
Benchmark                                                   Mode  Cnt  Score   Error  Units
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.029 ± 0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.027 ±  0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.029 ± 0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.023 ±  0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.026 ±  0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.028 ± 0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.024 ±  0.001   s/op
```
after changes:

```
Benchmark                                                   Mode  Cnt  Score    Error  Units
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.011 ±  0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.008 ±  0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.010 ± 0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.012 ± 0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.013 ± 0.001   s/op
GetExpectedWithdrawalsBenchmark.profileExpectedWithdrawals    ss  200  0.008 ±  0.001   s/op
```
Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/capella/block/BlockProcessorCapella.java [hunk 1]
```diff
@@ -230,31 +231,39 @@ private List<Withdrawal> getExpectedWithdrawals(final BeaconStateCapella preStat
     final List<Withdrawal> expectedWithdrawals = new ArrayList<>();
     final WithdrawalSchema withdrawalSchema = schemaDefinitionsCapella.getWithdrawalSchema();
     final UInt64 epoch = miscHelpers.computeEpochAtSlot(preState.getSlot());
-    final int validatorCount = preState.getValidators().size();
+    final SszList<Validator> validators = preState.getValidators();
+    final SszUInt64List balances = preState.getBalances();
+    final int validatorCount = validators.size();
+    final int maxValidatorsPerPayload = specConfigCapella.getMaxWithdrawalsPerPayload();
```
Comment (ajsutton) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/capella/block/BlockProcessorCapella.java]: nit: I'd call this `maxWithdrawalsPerPayload`

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/capella/block/BlockProcessorCapella.java [hunk 2]
```diff
@@ -230,31 +231,39 @@ private List<Withdrawal> getExpectedWithdrawals(final BeaconStateCapella preStat
...
-        expectedWithdrawals.add(
-            withdrawalSchema.create(
-                withdrawalIndex,
-                UInt64.valueOf(validatorIndex),
-                new Bytes20(validator.getWithdrawalCredentials().slice(12)),
-                balance.minus(specConfig.getMaxEffectiveBalance())));
-        withdrawalIndex = withdrawalIndex.increment();
+
+      final Validator validator = validators.get(validatorIndex);
+      if (predicates.hasEth1WithdrawalCredential(validator)) {
```
Comment (ajsutton) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/capella/block/BlockProcessorCapella.java:245]: Invert this to `if (!hasEth1WithdrawalCredentials) { continue; }` ?

**Review by ajsutton — APPROVED** [review]
LGTM.

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/capella/block/BlockProcessorCapella.java (hunk 2 above)
Comment (rolfyone) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/capella/block/BlockProcessorCapella.java:245]: can't continue, need to update the reference at the end of the loop...

**Pull request merged**